File: Data Structures & Algorithms/decode-ways/submission-4.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if len(s) == 0: return 0
        if s[0] == "0": return 0
        
        i = 0
        visited = {}

        def dfs( i: int) -> int:
            if i == len(s): return 1
            if s[i] == "0": return 0
            if i in visited: return visited[i]

            branchSum = 0
            
            branchSum += dfs(i+1)
            if i+1 < len(s) and int(s[i]+s[i+1]) <= 26:
                branchSum += dfs(i+2)

            visited[i] = branchSum
            return branchSum

        return dfs(0)

    


        

File: Data Structures & Algorithms/coin-change/submission-4.swift
class Solution {
    func coinChange(_ coins: [Int], _ amount: Int) -> Int {
        guard amount != 0 else { return 0 }
        var cache: [Int: Double] = [:]
        
        func dfs(_ currAmt: Int) -> Double {
            guard currAmt <= amount else { return Double.infinity }
            guard currAmt != amount else { return 0 }
            if cache[currAmt] != nil { return cache[currAmt]! }
            
            var ret: [Double] = []
            for c in coins{
                ret.append( 1+dfs(currAmt+c) )
            }
            let ans = ret.min()!
            
            cache[currAmt] = ans
            return ans
        }
        var ret: [Double] = []
        for c in coins{
            ret.append( 1+dfs(c) )
        }

        if ret.min()! == Double.infinity {
            return -1
        } else {
            return Int(ret.min()!)
        }
            

    }
}
